src/trait_group.cpp

```diff
@@ -0,0 +1,333 @@
+    }
+}
+
+bool Trait_group_creator::remove_trait( const trait_id &tid )
+{
+    Trait_creation_data *tcd = mutation_branch::get_group( id );
+
+    if( tcd ) {
+        tcd->remove_trait( tid );
+    }
+    return false;
+}
+
+bool Trait_group_creator::has_trait( const trait_id &tid ) const
+{
+    Trait_creation_data *tcd = mutation_branch::get_group( id );
+
+    return tcd && tcd->has_trait( tid );
+}
+
+void Trait_group::add_trait_entry( const trait_id &tid, int probability )
+{
+    std::unique_ptr<Trait_creation_data> ptr( new Single_trait_creator( tid, probability ) );
+    add_entry( ptr );
+}
+
+void Trait_group::add_group_entry( const Trait_group_tag &gid, int probability )
+{
+    std::unique_ptr<Trait_creation_data> ptr( new Trait_group_creator( gid, probability ) );
+    add_entry( ptr );
+}
+
+void Trait_group::check_consistency() const
+{
+    for( const auto &creator : creators ) {
+        creator->check_consistency();
+    }
+}
+
+bool Trait_group::remove_trait( const trait_id &tid )
+{
+    for( auto it = creators.begin(); it != creators.end(); ) {
+        if( ( *it )->remove_trait( tid ) ) {
+            sum_prob -= ( *it )->probability;
+            delete *it;
+            it = creators.erase( it );
+        } else {
+            ++it;
+        }
+    }
+    return creators.empty();
+}
+
+bool Trait_group::has_trait( const trait_id &tid ) const
+{
+    for( const auto &creator : creators ) {
+        if( creator->has_trait( tid ) ) {
+            return true;
+        }
+    }
+    return false;
+}
+
+Trait_group_collection::Trait_group_collection( int probability )
+    : Trait_group( probability )
+{
+    if( probability <= 0 || probability > 100 ) {
+        debugmsg( "Probability %d out of range", probability );
+    }
+};
+
+Trait_list Trait_group_collection::create( RecursionList &rec ) const
+{
+    Trait_list result;
+    for( const auto &creator : creators ) {
+        if( rng( 0, 99 ) >= ( creator )->probability ) {
+            continue;
+        }
+        Trait_list tmp = ( creator )->create( rec );
+        result.insert( result.end(), tmp.begin(), tmp.end() );
+    }
+    return result;
+}
+
+void Trait_group_collection::add_entry( std::unique_ptr<Trait_creation_data> &ptr )
+{
+    assert( ptr.get() != nullptr );
+    if( ptr->probability <= 0 ) {
+        return;
+    }
+
+    ptr->probability = std::min( 100, ptr->probability );
+
+    creators.push_back( ptr.get() );
+    ptr.release();
+}
+
+void Trait_group_distribution::add_entry( std::unique_ptr<Trait_creation_data> &ptr )
+{
+    assert( ptr.get() != nullptr );
+    if( ptr->probability <= 0 ) {
+        return;
+    }
+
+    sum_prob += ptr->probability;
+
+    creators.push_back( ptr.get() );
+    ptr.release();
+}
+
+Trait_list Trait_group_distribution::create( RecursionList &rec ) const
+{
+    Trait_list result;
+    int p = rng( 0, sum_prob - 1 );
+    for( const auto &creator : creators ) {
+        p -= ( creator )->probability;
```

Comment:
The brackets are unnecessary.

Reply:
This is a bit of stylistic feedback isn't it? Is this something covered in a style guide regarding brackets for single statement blocks?

Reply:
I think he means the parentheses around creator, ```( creator )->probability;```, brackets are mandatory.

Reply:
This is the style that I saw frequently in `item_factory.cpp` and `item_group.cpp`, I thought it was kind of weird too. I'll fix those in the new PR I split off from this.